### PR TITLE
ramips: add TP-Link RE200v3 support

### PR DESCRIPTION
**Pull request by misanthropos**
based on support for RE200v2
openwrt-ramips-mt76x8-tplink_re200-v3-squashfs-factory.bin can be installed via web-interface

### CODE REVIEW

File: target/linux/ramips/mt76x8/base-files/etc/board.d/02_network [hunk 1]
```diff
@@ -19,6 +19,7 @@ ramips_setup_interfaces()
 	ravpower,rp-wd009|\
 	tama,w06|\
 	tplink,re200-v2|\
+        tplink,re200-v3|\
```
Comment (adrianschmutzler) [inline target/linux/ramips/mt76x8/base-files/etc/board.d/02_network:22]: Indent.

**Comment by adrianschmutzler** [comment]
This has formal issues. Please add target prefix to commit title as well, add Specifications and Flashing instructions (e.g. [1]), and add a Signed-off-by.

[1] https://github.com/openwrt/openwrt/commit/6d6f36ae787c3b339b16c1d345a90faa1d9f4011

File: tools/firmware-utils/src/tplink-safeloader.c [hunk 2]
```diff
@@ -1562,6 +1562,49 @@ static struct device_info boards[] = {
...
+			"{product_name:RE200,product_ver:3.0.0,special_id:42520000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:43410000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:45530000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:45550000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:49440000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:4a500000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:4b520000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:52550000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:54570000}\n"
+			"{product_name:RE200,product_ver:3.0.0,special_id:55530000}\n",
```
Comment (adrianschmutzler) [inline tools/firmware-utils/src/tplink-safeloader.c:1583]: Have you properly extracted these, or just replaced "2.0.0" by "3.0.0" from v2?

Comment (blocktrron) [reply tools/firmware-utils/src/tplink-safeloader.c:1583]: EU Factory Image contains
```
strings re200v3_eu-up-ver1-1-0-P1\[20190731-rel41250\].bin | grep "special"
{product_name:RE200,product_ver:3.0.0,special_id:00000000}
{product_name:RE200,product_ver:3.0.0,special_id:45550000}
{product_name:RE200,product_ver:3.0.0,special_id:4A500000}
{product_name:RE200,product_ver:3.0.0,special_id:4B520000}
{product_name:RE200,product_ver:3.0.0,special_id:43410000}
{product_name:RE200,product_ver:3.0.0,special_id:41550000}
{product_name:RE200,product_ver:3.0.0,special_id:42520000}
{product_name:RE200,product_ver:3.0.0,special_id:55530000}
{product_name:RE200,product_ver:3.0.0,special_id:41520000}
{product_name:RE200,product_ver:3.0.0,special_id:52550000}
{product_name:RE200,product_ver:3.0.0,special_id:54570000}
{product_name:RE200,product_ver:3.0.0,special_id:45530000}
{product_name:RE200,product_ver:3.0.0,special_id:49440000}
{product_name:RE200,product_ver:3.0.0,special_id:45470000}
```

Comment (misanthropos) [reply tools/firmware-utils/src/tplink-safeloader.c:1583]: Does order matter? I copied the v2 declarations and compared them to the re200v3_eu-up-ver1-1-0-P1\[20190731-rel41250\].bin. They are the same except order. I can of course use the order coming with the bin file.

**Comment by adrianschmutzler** [comment]
> The MAC address assignment matches stock firmware, i.e.:
> LAN : *:0D
> 2.4G: *:0E
> 5G  : *:0F

Please don't try to trick me. The probability for your device to have the same last byte for the MAC address as v2 is 0.4 % if those are distributed evenly.

**Comment by misanthropos** [comment]
I did not try to trick anyone. It is a misunderstanding because the sentence ended with e.g.: and the MAC addresses only differ on the last tuple.

I am only assuming it has a UART header on PCB (57600 8n1) since it has the same chipset and form - I would have to pry it open to verify it actually does.

**Comment by adrianschmutzler** [comment]
> because the sentence ended with e.g.: 

The sentence ended with "i.e.", which means "specifically", or "which are", not "for example".

Anyway, that block testifies that the MAC address assignment matches stock firmware, so that somebody looking at it will know that you checked the assignment on vendor firmware. So, while for other data it might be fine to _assume_ that it's the same as v2, and just add it, in this case you specifically state that you _don't_ assume, but you have _checked_.

So, have you checked?
(Frequently, people go the painful way of trying to actually get the information now; however, I only ask a question for the moment, so just answering it might be cheaper ...)

> I am only assuming it has a UART header on PCB (57600 8n1) since it has the same chipset and form

Well, you have two options with that: 1. Find out the truth or 2. don't tell something as fact that isn't one.

Same is true for the instructions about serial console and flashing back to stock.

Just be honest about what you know and what you don't know. Otherwise, you might create a lot of useless work for people taking your assumptions for facts.

**Comment by adrianschmutzler** [comment]
Content-wise this looks fine BTW.

**Comment by misanthropos** [comment]
I have checked the MAC which is printed on the case is the same in OpenWRT for the LAN interface. They are the same.

File: tools/firmware-utils/src/tplink-safeloader.c (hunk 2 above)
Comment (blocktrron) [reply tools/firmware-utils/src/tplink-safeloader.c]: There were some missing IDs when compared with the v2. Order does not matter for the factory firmware, however i would prepare them to be ordered here.

**Comment by adrianschmutzler** [comment]
> I have checked the MAC which is printed on the case is the same in OpenWRT for the LAN interface. They are the same.

Okay. But that's not an answer to my question whether you have checked the MAC address assignment for LAN, WAN, Wifi on vendor firmware.

**Comment by misanthropos** [comment]
OK. I have not checked the MAC on the vendor firmware. I would have to revert the device to run with the vendor firmware to do that.

**Comment by adrianschmutzler** [comment]
Thanks, I've pulled this into my staging tree, sorted the support list and fine-tuned the commit description based on my desires.

Note that the discussion was not meant to bug you, but is necessary to prevent "false information" stated as fact being passed on to future developers looking at this.

**Pull request closed without merging**